isofit/radiative_transfer/radiative_transfer.py

```diff
@@ -325,18 +329,51 @@ def drdn_dRT(
         elif self.glint_model:
             L_down_transmitted = self.get_L_down_transmitted(x_RT, geom)
 
-            E_dd = (self.solar_irr * self.coszen) / np.pi * r["t_down_dir"]
-            E_ds = (self.solar_irr * self.coszen) / np.pi * r["t_down_dif"]
-            E_d = E_dd + E_ds
-            L_sky = x_surface[-2] * E_dd + x_surface[-1] * E_ds
+            t_down_dir = r["transm_down_dir"]  # downward direct transmittance
+            t_down_dif = r["transm_down_dif"]  # downward diffuse transmittance
+            t_down_total = t_down_dir + t_down_dif  # downward total transmittance
+            t_total_up = (
+                r["transm_up_dif"] + r["transm_up_dir"]
+            )  # total upward transmittance
+
+            L_sky = x_surface[-2] * t_down_dir + x_surface[-1] * t_down_dif
 
-            rho_ls = 0.02  # fresnel reflectance factor (approx. 0.02 for nadir view)
-            glint = rho_ls * (L_sky / E_d)
+            rho_ls = self.fresnel_rf(
+                geom.observer_zenith
+            )  # fresnel reflectance factor (approx. 0.02 for nadir view)
+            glint = rho_ls * (L_sky / t_down_total)
 
             drho_scaled_for_multiscattering_drfl = (
                 1.0 / (1 - r["sphalb"] * (rfl + glint)) ** 2
             )
-            drdn_drfl = L_down_transmitted * drho_scaled_for_multiscattering_drfl
+            drdn_drfl = (
+                t_total_up * L_down_transmitted * drho_scaled_for_multiscattering_drfl
+            )
+
+            # Basic formulation (below) does not include the derivative of the radiance w.r.t. other surface states, not just the reflectance
+            # There's probably a better fix for that overall, just trying to fix it for glint for now
+            drdn_dgdd = (
+                (self.solar_irr * self.coszen / np.pi)
+                * t_total_up
+                * t_down_dir
+                * drho_scaled_for_multiscattering_drfl
+            )
+            drdn_dgdsf = (
+                (self.solar_irr * self.coszen / np.pi)
+                * t_total_up
+                * t_down_dif
+                * drho_scaled_for_multiscattering_drfl
+            )
+
+            drdn_dLs = r["transm_up_dir"] + r["transm_up_dif"]
+            K_surface = (
+                drdn_drfl[:, np.newaxis] * drfl_dsurface
+                + drdn_dLs[:, np.newaxis] * dLs_dsurface
+            )
+            K_surface[:, -2] = drdn_dgdd
+            K_surface[:, -1] = drdn_dgdsf
```

Comment:
I think this can be handled generally for additional surface state elements as is done for RT elements in 296-301 above. Most critically, we need to avoid the hard coding of derivative positions.

Reply:
Good thought! I'll develop something more general, also wrt #524.

Reply:
To push the new glint model forward, we defer the implementation of general derivatives for additional surface state elements to a subsequent PR. The needs update is caught in #554.